File: src/main/java/mods/eln/sim/TimeRemover.java
package mods.eln.sim;

import mods.eln.Eln;
import mods.eln.Vars;

public class TimeRemover implements IProcess {

    ITimeRemoverObserver observer;
    double timeout = 0;

    public TimeRemover(ITimeRemoverObserver observer) {
        this.observer = observer;
    }

    public void setTimeout(double timeout) {
        if (this.timeout <= 0) {
            observer.timeRemoverAdd();
            Vars.simulator.addSlowProcess(this);
        }
        this.timeout = timeout;
    }

    @Override
    public void process(double time) {
        if (isArmed()) {
            timeout -= time;
            if (timeout <= 0) {
                shot();
            }
        }
    }

    public boolean isArmed() {
        return timeout > 0;
    }

    public void shot() {
        timeout = 0;
        observer.timeRemoverRemove();
        Vars.simulator.removeSlowProcess(this);
    }
}

File: src/main/java/mods/eln/sim/TimeRemover.kt
package mods.eln.sim

import mods.eln.Eln

class TimeRemover(internal var observer: ITimeRemoverObserver) : IProcess {
    internal var timeout = 0.0

    fun getIsArmed(): Boolean = timeout > 0

    fun setTimeout(timeout: Double) {
        if (this.timeout <= 0) {
            observer.timeRemoverAdd()
            Eln.simulator.addProcess(ProcessType.SlowProcess, this)
        }
        this.timeout = timeout
    }

    override fun process(time: Double) {
        if (getIsArmed()) {
            timeout -= time
            if (timeout <= 0) {
                shot()
            }
        }
    }

    fun shot() {
        timeout = 0.0
        observer.timeRemoverRemove()
        Eln.simulator.removeProcess(ProcessType.SlowProcess, this)
    }
}
